grammar.js

```diff
@@ -336,9 +359,62 @@ module.exports = grammar({
     // expressions
 
     _expression: $ => choice(
-      $._literal
+      $.identifier_name,
+      $._literal,
+      $.ternary_expression,
+      $.binary_expression,
+      $.unary_expression
     ),
 
+    ternary_expression: $ => prec.right(PREC.COND, seq(
+      $._expression, '?', $._expression, ':', $._expression
+    )),
+
+    binary_expression: $ => choice(
+      ...[
+        ['&&', PREC.LOGAND],
+        ['||', PREC.LOGOR],
+        ['>>', PREC.SHIFT],
+        ['<<', PREC.SHIFT],
+        ['&', PREC.AND],
+        ['^', PREC.OR],
+        ['|', PREC.OR],
+        ['+', PREC.ADD],
+        ['-', PREC.ADD],
+        ['*', PREC.MULT],
+        ['/', PREC.MULT],
+        ['%', PREC.MULT],
+        ['<', PREC.REL],
+        ['<=', PREC.REL],
+        ['==', PREC.EQUAL],
+        ['!=', PREC.EQUAL],
+        ['>=', PREC.REL],
+        ['>', PREC.REL],
+      ].map(([operator, precedence]) =>
+        prec.left(precedence, seq($._expression, operator, $._expression))
+      )
+    ),
+
+    unary_expression: $ => prec.right(PREC.UNARY, choice(
+      ...[
+        '!',
+        '~',
+        '-',
+        '+',
+        '--',
+        '++'
+      ].map(operator => seq(operator, $._expression))
+      .concat([
+        'typeof',
+        'sizeof'
+      ].map(operator => seq(operator, '(', $._expression, ')'))))),
```

Comment:
This will mean that the `unary_expression` node will have a varying number of children. Usually, it will just have two children (the operator and the operand expression), but in the case of `typeof` and `sizeof`, it will have four children (operator, left-paren, operand expression, and right-paren).

It might make the tree a little more uniform and easy to analyze if we treated `typeof` and `sizeof` just like any other unary operator. We could handle the parentheses independently via a `parenthesized_expression` rule which we'll eventually need anyway.

```js
parenthesized_expression: $ => seq('(', $._expression, ')'),
```

Reply:
That makes sense, and parenthesis are always legal so this doesn't break anything =]

Reply:
If we do want to enforce that parentheses are used with `sizeof` and `typeof`, we could keep your basic structure, but replace `seq(operator, '(', $._expression, ')')` with `seq(operator, $.parenthesized_expression)`.

I'd be ok with either approach. In general, I'm ok w/ parsing a *superset* of the language in question (e.g. allowing things like `typeof foo`) if it makes the syntax tree or the grammar implementation simpler.

Reply:
I like going for the superset approach, keeps things simple. And if this becomes an issue, someone will open a bug, and we already know the solution.